This notebook is used for only knowing and ANALYSING and VISUALIZING the data 

# Analization

In [134]:
import pandas as pd
import plotly as py
import numpy as np
import matplotlib.pyplot as plt
import plotly.graph_objects as go
import seaborn as sns              
import cufflinks as cf
from plotly.offline import iplot 
import plotly.express as px
py.offline.init_notebook_mode(connected=True)      #offline mode jupyter notebook will be connected
cf.go_offline()

agency_id                                       :Unique identifier for the agency
toptier_code	                                :Code for top-level federal agency classification
abbreviation	                                :Acronym of the agency name
agency_name	                                    :Full name of the federal agency
congressional_justification_url	                :Link to agency’s budget request document submitted to Congress
active_fy	                                    :Current fiscal year(financial year) for which data is reported
active_fq	                                    :Current fiscal quarter
outlay_amount	                                :Actual amount spent by the agency
obligated_amount                                :Funds committed by the agency but not yet spent
budget_authority_amount                         :Authorized budget for the agency
current_total_budget_authority_amount           :Total budget authority including all adjustments
percentage_of_total_budget_authority            :Agency’s share of the total government budget
agency_slug	                                    :URL-friendly version of the agency name used in web links

In [2]:
df = pd.read_csv("Cleaned_USASpendings.csv")

In [ ]:
df.describe()
#gives statistical analysis in one go

Unnamed: 0    agency_id  toptier_code  active_fy  active_fq  \
count  110.000000   110.000000    110.000000      110.0      110.0   
mean    54.500000   927.090909    483.727273     2025.0        3.0   
std     31.898276   389.814577   1198.493496        0.0        0.0   
min      0.000000    11.000000      5.000000     2025.0        3.0   
25%     27.250000   661.250000     70.250000     2025.0        3.0   
50%     54.500000  1096.500000    322.000000     2025.0        3.0   
75%     81.750000  1157.750000    462.000000     2025.0        3.0   
max    109.000000  1546.000000   9553.000000     2025.0        3.0   

       outlay_amount  obligated_amount  budget_authority_amount  \
count   1.100000e+02      1.100000e+02             1.100000e+02   
mean    5.839828e+10      6.029031e+10             1.255512e+11   
std     2.576655e+11      2.662548e+11             5.488774e+11   
min     0.000000e+00      0.000000e+00            -9.997344e+09   
25%     7.860313e+06      8.325629e+06             1.873225e+07   
50%     5.624689e+07      5.878291e+07             1.709333e+08   
75%     1.033160e+09      6.963359e+08             3.184529e+09   
max     1.826817e+12      1.934489e+12             4.253024e+12   

       current_total_budget_authority_amount  \
count                           1.100000e+02   
mean                            1.419547e+13   
std                             0.000000e+00   
min                             1.419547e+13   
25%                             1.419547e+13   
50%                             1.419547e+13   
75%                             1.419547e+13   
max                             1.419547e+13   

       percentage_of_total_budget_authority  
count                            110.000000  
mean                               0.008844  
std                                0.038666  
min                               -0.000704  
25%                                0.000001  
50%                                0.000012  
75%                                0.000224  
max                                0.299604

here we can see min amount of "budget_authority_amount" is negative, which is not possible, that amount must be positive 

In [33]:
wrongValue = df['budget_authority_amount'].idxmin()
wrongValue

46

In [37]:
df['budget_authority_amount'][wrongValue] = abs(df['budget_authority_amount'].min())


C:\Users\prithvirajp\AppData\Local\Temp\ipykernel_20424\830575591.py:1: FutureWarning:

ChainedAssignmentError: behaviour will change in pandas 3.0!
You are setting values through chained assignment. Currently this works in certain cases, but when using Copy-on-Write (which will become the default behaviour in pandas 3.0) this will never work to update the original DataFrame or Series, because the intermediate object on which we are setting values will behave as a copy.
A typical example is when you are setting values in a column of a DataFrame, like:

df["col"][row_indexer] = value

Use `df.loc[row_indexer, "col"] = values` instead, to perform the assignment in a single step and ensure this keeps updating the original `df`.

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy




In [43]:
df['budget_authority_amount'][wrongValue]

np.float64(9997343652.24)

Now the value is changed.

In [44]:
df.describe()

Unnamed: 0    agency_id  toptier_code  active_fy  active_fq  \
count  110.000000   110.000000    110.000000      110.0      110.0   
mean    54.500000   927.090909    483.727273     2025.0        3.0   
std     31.898276   389.814577   1198.493496        0.0        0.0   
min      0.000000    11.000000      5.000000     2025.0        3.0   
25%     27.250000   661.250000     70.250000     2025.0        3.0   
50%     54.500000  1096.500000    322.000000     2025.0        3.0   
75%     81.750000  1157.750000    462.000000     2025.0        3.0   
max    109.000000  1546.000000   9553.000000     2025.0        3.0   

       outlay_amount  obligated_amount  budget_authority_amount  \
count   1.100000e+02      1.100000e+02             1.100000e+02   
mean    5.839828e+10      6.029031e+10             1.257330e+11   
std     2.576655e+11      2.662548e+11             5.488354e+11   
min     0.000000e+00      0.000000e+00             0.000000e+00   
25%     7.860313e+06      8.325629e+06             1.932601e+07   
50%     5.624689e+07      5.878291e+07             1.795654e+08   
75%     1.033160e+09      6.963359e+08             3.395071e+09   
max     1.826817e+12      1.934489e+12             4.253024e+12   

       current_total_budget_authority_amount  \
count                           1.100000e+02   
mean                            1.419547e+13   
std                             0.000000e+00   
min                             1.419547e+13   
25%                             1.419547e+13   
50%                             1.419547e+13   
75%                             1.419547e+13   
max                             1.419547e+13   

       percentage_of_total_budget_authority  
count                            110.000000  
mean                               0.008844  
std                                0.038666  
min                               -0.000704  
25%                                0.000001  
50%                                0.000012  
75%                                0.000224  
max                                0.299604

# Visualization

## 1. Top 10 Agencies with most and least Outlay Amount

In [ ]:
mostAMT = df.sort_values(by='outlay_amount',ascending=False)
leastAMT = df.sort_values(by='outlay_amount').head(10)

In [175]:
mostAMT = mostAMT.iloc[:10]

In [179]:
fig1 = px.bar(mostAMT,x= 'abbreviation',y='outlay_amount',title="Most Spent on an Agency",
             color='percentage_of_total_budget_authority',
             hover_data=['agency_id','agency_name','toptier_code'],
             barmode='group',
             height=500)
fig1.show()


### ***Insights***:-
**"Most Spent on an Agency"**:

1. **HHS (Health and Human Services)**  
   - **Outlay Amount**: Over $1.82 trillion    
   - HHS receives the largest portion of the federal budget, likely due to major healthcare programs like Medicare and Medicaid.

2. **TREAS (Department of the Treasury)**  
   - **Outlay Amount**: Over$1.34 trillion  
   - Significant spending likely includes interest on national debt and financial programs.

3. **SSA (Social Security Administration)**  
   - **Outlay Amount**: Over $1.14 trillion  
   - Reflects large expenditures on Social Security benefits.

4. **DOD (Department of Defense)**  
   - **Outlay Amount**: Slightly below $1 trillion  
   - High spending on military operations, personnel, and defense infrastructure.

5. **VA (Department of Veterans Affairs)**  
  - **Outlay Amount**: Around $278 billion 
  - Focused on veteran healthcare and benefits.

6. **USDA, OPM, DHS, DOT, HUD**  
  - **Outlay Amounts**: Significantly lower than top agencies  
  - These agencies have more specialized or narrower scopes of operation.

## 2. Top 10 agencies with Highest Obligation Amount

In [177]:
highObli = df.sort_values('obligated_amount',ascending=False).head(10)

In [180]:
fig2 = px.bar(highObli,x='abbreviation',y='obligated_amount',
              color='percentage_of_total_budget_authority',
              hover_data=['agency_id','agency_name','toptier_code'],
              )
fig2.show()

### ***Insights***:-
**"Most Obligated money on an Agency"**:

1. **HHS (Health and Human Services)**  
   - **Outlay Amount**: Over $1.93 trillion    

2. **TREAS (Department of the Treasury)**  
   - **Outlay Amount**: Over$1.34 trillion  

3. **SSA (Social Security Administration)**  
   - **Outlay Amount**: Over $1.16 trillion  

4. **DOD (Department of Defense)**  
   - **Outlay Amount**: Slightly above $1 trillion  

5. **VA (Department of Veterans Affairs)**  
  - **Outlay Amount**: Around $268 billion 

## 3. Top 10 agencies with highest Budget Authority Amount

In [178]:
highBudgetAuhtorityAmount = df.sort_values('budget_authority_amount',ascending=False).head(10)

In [181]:
fig3 = px.bar(highBudgetAuhtorityAmount,x='abbreviation',y='budget_authority_amount',
              color='percentage_of_total_budget_authority',
              hover_data=['agency_id','agency_name','toptier_code'],
              )
fig3.show()

## 4. Analysis of all the parameter in Histogram format

In [183]:
fig = go.Figure(data=[
    go.Bar(name='Outlay', x=mostAMT['abbreviation'], y=mostAMT['outlay_amount']),
    go.Bar(name='Obligated', x=highObli['abbreviation'], y=highObli['obligated_amount']),
    go.Bar(name='Authorized',x=highBudgetAuhtorityAmount['abbreviation'],y=highBudgetAuhtorityAmount['budget_authority_amount']),
    #go.Bar(hoverinfo=['agency_id','agency_name','toptier_code'])
])
fig.update_layout(barmode='group')
fig.show()



### 🔍 **Key Definitions**
- **Outlay**: Actual spending during the fiscal year.
- **Obligated**: Funds committed to be spent.
- **Authorized**: Maximum amount approved by legislation.

---

### 🏆 **Top Agencies by Financial Scale**

#### 1. **TREAS (Department of the Treasury)**
- **Outlay**: ~\$2T  
- **Obligated**: ~\$2.5T  
- **Authorized**: ~\$4.5T  
- **Insight**: Largest financial footprint, likely due to debt servicing, tax refunds, and financial programs.

#### 2. **HHS (Health and Human Services)**
- **Outlay**: ~\$1T  
- **Obligated**: ~\$1.5T  
- **Authorized**: ~\$2.5T  
- **Insight**: Major healthcare programs drive high spending and commitments.

#### 3. **SSA (Social Security Administration)**
- **Outlay**: ~\$1T  
- **Obligated**: ~\$1.25T  
- **Authorized**: ~\$1.75T  
- **Insight**: Reflects consistent funding for Social Security benefits.

#### 4. **DOD (Department of Defense)**
- **Outlay**: <\$0.5T  
- **Obligated**: <\$0.75T  
- **Authorized**: >\$2.25T  
- **Insight**: High authorization suggests readiness for large-scale operations, though actual spending is lower.

---

### 📉 **Lower Spending Agencies**
All below ~\$0.25T in all categories:
- **VA (Veterans Affairs)**
- **USDA (Agriculture)**
- **OPM (Personnel Management)**
- **DHS (Homeland Security)**
- **DOT (Transportation)**
- **HUD (Housing and Urban Development)**
- **DOE (Energy)** – notably the lowest, <\$0.05T

---

### 📊 **Insights & Observations**
- **TREAS and HHS** dominate in all three financial categories.
- **DOD** has a large gap between authorized and actual outlay, indicating strategic flexibility.
- **SSA** maintains a balanced and predictable financial profile.
- **Smaller agencies** operate with relatively modest budgets, reflecting narrower scopes or specialized missions.

## 5. Analysis of Percentage of Total Budget using Pie Chart

In [171]:
percentage = df.sort_values(by='percentage_of_total_budget_authority',ascending=False).head(15)


In [182]:
fig4 = px.pie(percentage,values=percentage['percentage_of_total_budget_authority'],names=percentage['abbreviation'],
              title="Percentage share of each agency",
              color_discrete_sequence=px.colors.sequential.RdBu,
              hover_data=['agency_id','agency_name','outlay_amount'])
fig4.show()